euphoria-spark/src/main/java/cz/seznam/euphoria/spark/JoinTranslator.java

```diff
@@ -0,0 +1,128 @@
+        joined = leftPair.fullOuterJoin(rightPair, partitioning.getNumPartitions())
+            .mapToPair(t -> new Tuple2<>(t._1, new Tuple2<>(t._2._1.orNull(), t._2._2.orNull())));
+      } else {
+        joined = leftPair.join(rightPair, partitioning.getNumPartitions());
+      }
+    } else {
+      // ~ pass partitioner implementation
+      if (operator.isOuter()) {
+        joined = leftPair.fullOuterJoin(rightPair, new PartitioningWrapper(partitioning))
+            .mapToPair(t -> new Tuple2<>(t._1, new Tuple2<>(t._2._1.orNull(), t._2._2.orNull())));
+      } else {
+        joined = leftPair.join(rightPair, new PartitioningWrapper(partitioning));
+      }
+    }
+
+    // @TODO ?????? I DON'T KNOW WHAT I'M DOING
+    final LazyAccumulatorProvider acc =
+        new LazyAccumulatorProvider(context.getAccumulatorFactory(), context.getSettings());
+
+    return joined.flatMap(new FlatMapFunctionWithContext<>((t, collector) -> {
+      // ~ both elements have exactly the same window
+      // we need to check for null values because of full outer join
+      final Window window = t._2._1 == null ? t._2._2.getWindow() : t._2._1.getWindow();
+      final long maxTimestamp = Math.max(
+          t._2._1 == null ? 0L : t._2._1.getTimestamp(),
+          t._2._2 == null ? 0L : t._2._2.getTimestamp());
+      collector.clear();
+      collector.setWindow(window);
+      operator.getJoiner().apply(
+          t._2._1 == null ? null : t._2._1.getElement(),
+          t._2._2 == null ? null : t._2._2.getElement(),
+          collector);
+      return Iterators.transform(collector.getOutputIterator(), e ->
+          new SparkElement(window, maxTimestamp, Pair.of(t._1.key(), e)));
+    }, acc));
+  }
+
+  /**
+   * Expand element to all windows
+   */
+  private static class KeyExtractor implements PairFlatMapFunction<SparkElement, KeyedWindow, SparkElement> {
+
+    private final UnaryFunction keyExtractor;
+    private final Windowing windowing;
+    private final boolean left;
+
+    KeyExtractor(UnaryFunction keyExtractor, Windowing windowing, boolean left) {
+      this.keyExtractor = keyExtractor;
+      this.windowing = windowing;
+      this.left = left;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public Iterator<Tuple2<KeyedWindow, SparkElement>> call(SparkElement se) throws Exception {
+      // ~ shit just got real ... @todo discuss operator test kit
+      final Iterable<Window> windows = windowing.assignWindowsToElement(new SparkElement(se.getWindow(),
+          se.getTimestamp(), left ? Either.left(se.getElement()) : Either.right(se.getElement())));
+      final List<Tuple2<KeyedWindow, SparkElement>> out = new ArrayList<>();
+      for (Window window : windows) {
+        out.add(new Tuple2<>(
+            new KeyedWindow<>(window, se.getTimestamp(), keyExtractor.apply(se.getElement())),
+            new SparkElement(window, se.getTimestamp(), se.getElement())));
+      }
+      return out.iterator();
```

Comment:
Here is unnecessary placing elements into new collection. For performance reasons it would be better to use something like `Iterators.transform`.